terminal/src/main/java/org/jline/terminal/impl/AbstractWindowsTerminal.java

```diff
@@ -176,13 +176,74 @@ public void close() throws IOException {
     }
 
-    protected String getEscapeSequence(short keyCode) {
+    final int ctrlFlag = 4;
+    final int altFlag = 2;
+    final int shiftFlag = 1;
+
+    protected String getEscapeSequenceFromConsoleInput(final int[] keyEvent) {
+        final int KEY_DOWN = 0;
+        final int KEY_CODE = 1;
+        final int KEY_CHAR = 2;
+        final int KEY_STAT = 3;
+        final int KEY_REPEAT = 4;
+        final int KEY_SCAN = 5;
+        final int altState = 0x0002 | 0x0001;
+        final int ctrlState = 0x0008 | 0x0004;
+        final int shiftState = 0x0010;
+        final boolean isCtrl = (keyEvent[KEY_STAT] & ctrlState) > 0;
+        final boolean isAlt = (keyEvent[KEY_STAT] & altState) > 0;
+        final boolean isShift = (keyEvent[KEY_STAT] & shiftState) > 0;
+        char ch = (char) keyEvent[KEY_CHAR];
+        short code = (short) keyEvent[KEY_CODE];
+        StringBuilder sb = new StringBuilder(32);
+
+        // key down event
+        if (keyEvent[KEY_DOWN] == 1 && ch != 3) {
+            if (isShift && ch == '\t') return getSequence(InfoCmp.Capability.key_btab);
+            final String keySeq = getEscapeSequence(code, (isCtrl ? ctrlFlag : 0) + (isAlt ? altFlag : 0) + (isShift ? shiftFlag : 0));
+            if (keySeq != null) return keySeq;
+            if (ch >= 32 && ch <= 126) { //Visible ANSI character
+                if (isAlt) sb.append("\033");
+                if (isCtrl) {
+                    sb.append(String.valueOf(ch == '?' ? ((char) 0x7f) : Character.toUpperCase(ch) & 0x1f));
+                } else {
+                    sb.append(ch);
+                }
+            } else if (ch > 0) {
+                sb.append(ch);
+            } else if (isCtrl && code >= 'A' && code <= 'Z') { //Handles the ctrl key events(uchar=0)
+                if (code >= 'A' && code <= 'Z') {
+                    ch = (char) (code - 0x40);
+                } else if (code == 191) {
+                    ch = 127;
+                }
+                if (ch > 0) {
+                    if (isAlt) sb.append("\033");
+                    sb.append(ch);
+                }
+            }
+        } else {// key up event
+            //In Win10 Ctrl+C doesn't have keydown event
+            if (ch == 3) return "\3";
+            // support ALT+NumPad input method
+            if (keyEvent[KEY_CODE] == 0x12 /*VK_MENU ALT key*/ && ch > 0)
+                sb.append(ch);  // no such combination in windows
+
+        }
+        return sb.toString();
+    }
+
+    protected String getEscapeSequence(short keyCode, int keyState) {
         // virtual keycodes: http://msdn.microsoft.com/en-us/library/windows/desktop/dd375731(v=vs.85).aspx
         // TODO: numpad keys, modifiers
         String escapeSequence = null;
+        String fmt = null;
         switch (keyCode) {
             case 0x08: // VK_BACK BackSpace
-                escapeSequence = getSequence(InfoCmp.Capability.key_backspace);
+                if ((keyState & altFlag) == 0)
```

Comment:
I'm not sure it makes sense any more to keep using the {{getSequence()}} calls if it only cover half the cases.  Imho, the code would be cleaner if using only constants.